File: src/cpp/1.two-sum.cpp
#include <vector>
#include <map>

/*
 * @lc app=leetcode id=1 lang=cpp
 *
 * [1] Two Sum
 * 
 * Time complexity: O(n)
 * Space complexity: O(n)
 */

// @lc code=start
class Solution
{
public:
    std::vector<int> twoSum(std::vector<int> &nums, int target)
    {
        std::map<int, int> seen;

        for (int i = 0; i < nums.size(); i++)
        {
            if (auto j = seen.find(target - nums[i]); j != seen.end())
            {
                std::vector<int> ans = {i, j->second};
                return ans;
            }

            seen.insert({nums[i], i});
        }

        return std::vector<int>();
    }
};
// @lc code=end

File: src/rust/1.two-sum.rs
/*
 * @lc app=leetcode id=1 lang=rust
 *
 * Time complexity: O(n)
 * Space complexity: O(n)
 */

// @lc code=start
use std::collections::HashMap;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut seen = HashMap::new();

        for i in 0..nums.len() {
            let rem = target - nums[i];

            if let Some(&j) = seen.get(&rem) {
                return vec![i as i32, j as i32];
            }

            seen.insert(nums[i], i);
        }

        return vec![];
    }
}
// @lc code=end